// ngx-collapse/esm5/lib/collapse-ctrl.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/collapse-ctrl.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, HostBinding, HostListener, Input } from '@angular/core';
import { CollapseContentDirective } from './collapse-content.directive';
// tslint:disable-next-line: directive-selector
var CollapseCtrlDirective = /** @class */ (function () {
    function CollapseCtrlDirective() {
        // @HostBinding('class.expand')
        this.expand = false;
    }
    Object.defineProperty(CollapseCtrlDirective.prototype, "contentRef", {
        set: /**
         * @param {?} v
         * @return {?}
         */
        function (v) {
            if (v instanceof CollapseContentDirective) {
                this.content = Array.of(v);
            }
            else if (Array.isArray(v)) {
                this.content = v.filter((/**
                 * @param {?} c
                 * @return {?}
                 */
                function (c) { return c instanceof CollapseContentDirective; }));
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CollapseCtrlDirective.prototype, "collapsed", {
        get: /**
         * @return {?}
         */
        function () {
            return !this.content.every((/**
             * @param {?} c
             * @return {?}
             */
            function (c) { return c.open; }));
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    CollapseCtrlDirective.prototype.onCLick = /**
     * @return {?}
     */
    function () {
        this.expand = !this.content.every((/**
         * @param {?} c
         * @return {?}
         */
        function (c) { return c.open; }));
        if (!this.content) {
            return;
        }
        this.content.forEach((/**
         * @param {?} c
         * @return {?}
         */
        function (c) { return c.toggle(); }));
    };
    /**
     * @param {?} c
     * @return {?}
     */
    CollapseCtrlDirective.prototype.hide = /**
     * @param {?} c
     * @return {?}
     */
    function (c) {
        c.hide();
    };
    /**
     * @param {?} c
     * @return {?}
     */
    CollapseCtrlDirective.prototype.show = /**
     * @param {?} c
     * @return {?}
     */
    function (c) {
        c.show();
    };
    CollapseCtrlDirective.decorators = [
        { type: Directive, args: [{ selector: '[collapse]' },] }
    ];
    /** @nocollapse */
    CollapseCtrlDirective.ctorParameters = function () { return []; };
    CollapseCtrlDirective.propDecorators = {
        contentRef: [{ type: Input, args: ['collapse',] }],
        collapsed: [{ type: HostBinding, args: ['class.collapsed',] }],
        onCLick: [{ type: HostListener, args: ['click',] }]
    };
    return CollapseCtrlDirective;
}());
export { CollapseCtrlDirective };
if (false) {
    /** @type {?} */
    CollapseCtrlDirective.prototype.content;
    /**
     * @type {?}
     * @private
     */
    CollapseCtrlDirective.prototype.expand;
}
