// ngx-collapse/esm5/lib/collapse-content.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/collapse-content.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, HostListener } from '@angular/core';
// tslint:disable-next-line: directive-selector
var CollapseContentDirective = /** @class */ (function () {
    function CollapseContentDirective(elementRef) {
        this.elementRef = elementRef;
        this.nextIsOpen = null;
        this._open = false;
    }
    /**
     * @private
     * @return {?}
     */
    CollapseContentDirective.prototype.getHTMLEL = /**
     * @private
     * @return {?}
     */
    function () {
        return this.elementRef.nativeElement;
    };
    /**
     * @return {?}
     */
    CollapseContentDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.getHTMLEL().classList.add('collapse');
    };
    /**
     * @return {?}
     */
    CollapseContentDirective.prototype.onTransitionComplete = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var element = this.getHTMLEL();
        if (this.nextIsOpen == true) {
            element.style.height = '';
            // this.collapsing = false;
            element.classList.remove('collapsing');
            element.classList.add('collapse');
            // this.collapse = true;
            element.classList.add('show');
        }
        else if (this.nextIsOpen == false) {
            // this.collapsing = false;
            element.classList.remove('collapsing');
            element.classList.add('collapse');
            // this.collapse = true;
        }
    };
    /**
     * @return {?}
     */
    CollapseContentDirective.prototype.hide = /**
     * @return {?}
     */
    function () {
        this.nextIsOpen = false;
        this._open = false;
        /** @type {?} */
        var element = this.getHTMLEL();
        // // remove collapse show
        element.classList.remove('collapse');
        element.classList.remove('show');
        /** @type {?} */
        var height = element.offsetHeight;
        element.style.height = height + "px";
        element.offsetHeight;
        // // collapsing and transition
        // this.collapsing = true;
        element.classList.add('collapsing');
        element.style.height = '';
    };
    /**
     * @return {?}
     */
    CollapseContentDirective.prototype.show = /**
     * @return {?}
     */
    function () {
        this.nextIsOpen = true;
        this._open = true;
        /** @type {?} */
        var element = this.getHTMLEL();
        // this.collapse = false;
        // this.collapsing = true;
        element.classList.remove('collapse');
        element.classList.add('collapsing');
        /** @type {?} */
        var height = element.scrollHeight;
        element.style.height = height + "px";
    };
    /**
     * @return {?}
     */
    CollapseContentDirective.prototype.toggle = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var element = this.getHTMLEL();
        element.classList.contains('show') ?
            this.hide() :
            this.show();
    };
    Object.defineProperty(CollapseContentDirective.prototype, "open", {
        get: /**
         * @return {?}
         */
        function () {
            return this._open || this.getHTMLEL().classList.contains('show');
        },
        enumerable: true,
        configurable: true
    });
    CollapseContentDirective.decorators = [
        { type: Directive, args: [{ selector: '[collapseContent]', exportAs: 'collapseContent' },] }
    ];
    /** @nocollapse */
    CollapseContentDirective.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    CollapseContentDirective.propDecorators = {
        onTransitionComplete: [{ type: HostListener, args: ['transitionend',] }]
    };
    return CollapseContentDirective;
}());
export { CollapseContentDirective };
if (false) {
    /**
     * @type {?}
     * @private
     */
    CollapseContentDirective.prototype.nextIsOpen;
    /**
     * @type {?}
     * @private
     */
    CollapseContentDirective.prototype._open;
    /**
     * @type {?}
     * @private
     */
    CollapseContentDirective.prototype.elementRef;
}
