// ngx-collapse/esm2015/lib/collapse-content.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/collapse-content.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, HostListener } from '@angular/core';
// tslint:disable-next-line: directive-selector
export class CollapseContentDirective {
    /**
     * @param {?} elementRef
     */
    constructor(elementRef) {
        this.elementRef = elementRef;
        this.nextIsOpen = null;
        this._open = false;
    }
    /**
     * @private
     * @return {?}
     */
    getHTMLEL() {
        return this.elementRef.nativeElement;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.getHTMLEL().classList.add('collapse');
    }
    /**
     * @return {?}
     */
    onTransitionComplete() {
        /** @type {?} */
        const element = this.getHTMLEL();
        if (this.nextIsOpen == true) {
            element.style.height = '';
            // this.collapsing = false;
            element.classList.remove('collapsing');
            element.classList.add('collapse');
            // this.collapse = true;
            element.classList.add('show');
        }
        else if (this.nextIsOpen == false) {
            // this.collapsing = false;
            element.classList.remove('collapsing');
            element.classList.add('collapse');
            // this.collapse = true;
        }
    }
    /**
     * @return {?}
     */
    hide() {
        this.nextIsOpen = false;
        this._open = false;
        /** @type {?} */
        const element = this.getHTMLEL();
        // // remove collapse show
        element.classList.remove('collapse');
        element.classList.remove('show');
        /** @type {?} */
        const height = element.offsetHeight;
        element.style.height = `${height}px`;
        element.offsetHeight;
        // // collapsing and transition
        // this.collapsing = true;
        element.classList.add('collapsing');
        element.style.height = '';
    }
    /**
     * @return {?}
     */
    show() {
        this.nextIsOpen = true;
        this._open = true;
        /** @type {?} */
        const element = this.getHTMLEL();
        // this.collapse = false;
        // this.collapsing = true;
        element.classList.remove('collapse');
        element.classList.add('collapsing');
        /** @type {?} */
        const height = element.scrollHeight;
        element.style.height = `${height}px`;
    }
    /**
     * @return {?}
     */
    toggle() {
        /** @type {?} */
        const element = this.getHTMLEL();
        element.classList.contains('show') ?
            this.hide() :
            this.show();
    }
    /**
     * @return {?}
     */
    get open() {
        return this._open || this.getHTMLEL().classList.contains('show');
    }
}
CollapseContentDirective.decorators = [
    { type: Directive, args: [{ selector: '[collapseContent]', exportAs: 'collapseContent' },] }
];
/** @nocollapse */
CollapseContentDirective.ctorParameters = () => [
    { type: ElementRef }
];
CollapseContentDirective.propDecorators = {
    onTransitionComplete: [{ type: HostListener, args: ['transitionend',] }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    CollapseContentDirective.prototype.nextIsOpen;
    /**
     * @type {?}
     * @private
     */
    CollapseContentDirective.prototype._open;
    /**
     * @type {?}
     * @private
     */
    CollapseContentDirective.prototype.elementRef;
}
