// ngx-collapse/esm2015/lib/collapse-ctrl.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/collapse-ctrl.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, HostBinding, HostListener, Input } from '@angular/core';
import { CollapseContentDirective } from './collapse-content.directive';
// tslint:disable-next-line: directive-selector
export class CollapseCtrlDirective {
    constructor() {
        // @HostBinding('class.expand')
        this.expand = false;
    }
    /**
     * @param {?} v
     * @return {?}
     */
    set contentRef(v) {
        if (v instanceof CollapseContentDirective) {
            this.content = Array.of(v);
        }
        else if (Array.isArray(v)) {
            this.content = v.filter((/**
             * @param {?} c
             * @return {?}
             */
            c => c instanceof CollapseContentDirective));
        }
    }
    /**
     * @return {?}
     */
    get collapsed() {
        return !this.content.every((/**
         * @param {?} c
         * @return {?}
         */
        c => c.open));
    }
    /**
     * @return {?}
     */
    onCLick() {
        this.expand = !this.content.every((/**
         * @param {?} c
         * @return {?}
         */
        c => c.open));
        if (!this.content) {
            return;
        }
        this.content.forEach((/**
         * @param {?} c
         * @return {?}
         */
        c => c.toggle()));
    }
    /**
     * @param {?} c
     * @return {?}
     */
    hide(c) {
        c.hide();
    }
    /**
     * @param {?} c
     * @return {?}
     */
    show(c) {
        c.show();
    }
}
CollapseCtrlDirective.decorators = [
    { type: Directive, args: [{ selector: '[collapse]' },] }
];
/** @nocollapse */
CollapseCtrlDirective.ctorParameters = () => [];
CollapseCtrlDirective.propDecorators = {
    contentRef: [{ type: Input, args: ['collapse',] }],
    collapsed: [{ type: HostBinding, args: ['class.collapsed',] }],
    onCLick: [{ type: HostListener, args: ['click',] }]
};
if (false) {
    /** @type {?} */
    CollapseCtrlDirective.prototype.content;
    /**
     * @type {?}
     * @private
     */
    CollapseCtrlDirective.prototype.expand;
}
